### ISSUE TITLE
Add more checks to system_check

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#155]
Currently we are only checking few things in system_check. We should extend it and check more things with more generic checks. We should check:

- If necessary settings are defined.
- If necessary settings have correct values.
- If RELEASE-VERSION exists.

More points might be added to this list.

**Comment by rwrzesien** [comment]
Old description as a reference:

Currently we are only checking few things in system_check. We should extend it and check more things with more generic checks. We should check:

- If necessary settings are defined.
- If necessary settings have correct values.
- If RELEASE-VERSION exists.
- If `CREATION_NEW_CHAIN_SEGMENT_TIME` is different than 0.
- If CONCENT_PUBLIC_KEY and CONCENT_PRIVATE_KEY are correct (length, type etc.).
- If all databases have `ATOMIC_REQUESTS` enabled (#279).

More points might be added to this list.

**Comment by cameel** [comment]
I like the idea. This is something repetitive that certainly can be improved.

1) Maybe use something like `check.Type`, `check.Required` instead of `SystemCheckType` and `SystemCheckRequired`?

2) Instead of `(GreaterThanOrEqualSystemCheck, [1])` you could do `(check.GreaterThanOrEqual, 1)` by assuming that the first element is always the check class and the rest of parameters are passed directly to `__init__()`.

    But maybe it would be better not to do anything complicated in `__init__()` and just create instances right away? They should not really have any local state except for their parameters.

3) `RequiredForConcentFeatureSystemCheck`: should support multiple features. And forcing the setting not to be defined when feature is disabled.

4) > One note - I don't like current way of setting system check error id like concent.E002. We only have few of those, but we already have some of them repeated, and there is no way to automatically prevent this. Typical case when this may happen is when two or more developers are defining system check erros in separate branches. Maybe we should rethink how this is done ?

    This is the default Django convention but we could change it for our own checks. For example to names similar to error codes.

5) > Do you see possibility to create tests for system checks in the same way ?

    We could have a test case class and special assertions for stuff that repeats in system checks.

**Comment by rwrzesien** [comment]
> 1. Maybe use something like check.Type, check.Required instead of SystemCheckType and SystemCheckRequired?

:+1: Sounds good, just to be sure you are saying that there should be module `check.py` which should be imported, and then we should use classes like `Type` from it ?

> 2. Instead of (GreaterThanOrEqualSystemCheck, [1]) you could do (check.GreaterThanOrEqual, 1) by assuming that the first element is always the check class and the rest of parameters are passed directly to __init__().

:+1: 

> 3. RequiredForConcentFeatureSystemCheck: should support multiple features. And forcing the setting not to be defined when feature is disabled.

:+1: Could be, what I have written was just a sample proposal.

> 4. This is the default Django convention but we could change it for our own checks. For example to names similar to error codes.

The only problem I have with it are numbers which can lead to issues, error codes should be fine.

**Comment by cameel** [comment]
> just to be sure you are saying that there should be module check.py which should be imported, and then we should use classes like Type from it

Yeah. Or `system_check`. The same way that `message` works in golem-messages.

> I agree with first paragraph, but about second one, wouldn't it allocate memory for whole process time if we do it this way ?

It would. Do you intend to store something big inside? I don't think that the difference between a class instance and a tuple is big enough to matter in this use case. If the class uses `__slots__` (and it should) an instance takes up 16 bytes with no fields, 56 bytes with one field, 64 with two, and so on. As long as `SYSTEM_CHECK_CONFIG` is a constant structure created at startup and not something that can grow at runtime I think it's not worth worrying about. My criteria for choosing the best solution would be what gives us better readability and less repetition.

> The question is where should this config be placed ?

Depends on whether you want something centralized like we have now or a system where you can place checks anywhere and register them like the current Django checks. I'm fine with either but I'd probably go with a centralized solution because none of our apps are really independent and having one place where we can find them all is convenient. After all we have settings in a central place too. I'd keep it in the current `system_check` module. But make sure that we don't run into problems with circular imports that way.

**Comment by rwrzesien** [comment]
> It would. Do you intend to store something big inside? 

Ok, I will update the example in description 

> Depends on whether you want something centralized like we have now or a system where you can place checks anywhere and register them like the current Django checks. I'm fine with either but I'd probably go with a centralized solution because none of our apps are really independent and having one place where we can find them all is convenient. After all we have settings in a central place too. I'd keep it in the current system_check module. But make sure that we don't run into problems with circular imports that way.

Yes, centralized is ok, I have rather meant physically. If this config should be placed in settings, or in some constants module, or in system_check directly. I think it is worth deciding before implementation starts.

**Comment by dybi** [comment]
@rwrzesien ,

from my understanding we could register one generic (master) `FullSystemCheck`, which would take the below config
```
SYSTEM_CHECK_CONFIG = {
    'MINIMUM_UPLOAD_RATE': [
        RequiredSystemCheck,    # generic
        (TypeSystemCheck, [int])    # generic, with list of parameters
        (GreaterThanOrEqualSystemCheck, [1])    # generic, with list of parameters
    ],
    'AVERAGE_BLOCK_TIME': [
        SCIBackendSetSystemCheck,    # specific for this setting, but might be reused too
        RequiredSystemCheck,    # generic
        (TypeSystemCheck, [int])    # generic, with list of parameters
        (GreaterThanSystemCheck, [0])    # generic, with list of parameters
    ],
}
```
and for each item use one of slave check functions (`RequiredSystemCheck`, `TypeSystemCheck`), etc. Then we could have simple test cases for slave checks, and some more sophisticated for registered master...
This way we could avoid inheritance and using too much abstraction.

@rwrzesien, @cameel thoughts?

**Comment by cameel** [comment]
@rwrzesien 
>  I have rather meant physically. If this config should be placed in settings, or in some constants module, or in system_check directly.

Definitely not in settings. It's something that should be hard-coded, not customized at deployment time.

As for `constants` vs `system_check`, I'd put it in the latter. I don't think there's any value it making it available as a generic global constant.

@dybi 
> Then we could have simple test cases for slave checks, and some more sophisticated for registered master...
> This way we could avoid inheritance and using too much abstraction.

Could you give a more concrete example? Do you mean a big "runner" for type checks and a set of simple functions and parameters for them?

**Comment by rwrzesien** [comment]
Updated description about where `SYSTEM_CHECK_CONFIG` should be placed and example showing that  system checks should be instantiated.
